### PR TITLE
flint 3.1.3

### PR DESCRIPTION
**Pull request by edgarcosta**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

<del>

I'm not sure what you would like me to do about:
> flint
>   * line 39, col 5: Formulae should be verified as having support for runtime hardware detection before using ENV.runtime_cpu_detection.
> Error: 1 problem in 1 formula detected.

</del>

depends on #166574

### CODE REVIEW

**Comment by SMillerDev** [comment]
Can you make disabling ARB a different PR? That should solve the issue I believe

**Comment by edgarcosta** [comment]
#166574 has been merged, so no longer issues with arb, but I still do not know how to address 
> Formulae should be verified as having support for runtime hardware detection before using ENV.runtime_cpu_detection.

File: Formula/f/flint.rb [hunk 1]
```diff
@@ -21,23 +21,31 @@ class Flint < Formula
...
+    # arch-specific code with runtime detection of capabilities:
+    # https://flintlib.org/doc/building.html
+    ENV.runtime_cpu_detection
+
+    system "./bootstrap.sh" if build.head?
+
     args = %W[
       --with-gmp=#{Formula["gmp"].prefix}
       --with-mpfr=#{Formula["mpfr"].prefix}
       --with-ntl=#{Formula["ntl"].prefix}
```
Comment (albinahlback) [inline Formula/f/flint.rb:45]: FLINT does not need NTL to install `NTL-interface.h`. Removing `--with-ntl` may reduce build-time and may reduce the complexity for installing.

However, you may still want FLINT to depend on NTL if you want to download NTL when downloading FLINT. `--with-ntl` is only used for checking, but perhaps there is some `make check` here I didn't see?

File: Formula/f/flint.rb [hunk 2]
```diff
@@ -1,13 +1,13 @@
...
+  url "https://github.com/flintlib/flint/releases/download/v3.1.1/flint-3.1.2.tar.gz"
+  sha256 "fdb3a431a37464834acff3bdc145f4fe8d0f951dd5327c4c6f93f4cbac5c2700"
   license "LGPL-3.0-or-later"
-  head "https://github.com/wbhart/flint2.git", branch: "trunk"
+  head "https://github.com/flintlib/flint.git", branch: "main"
 
   livecheck do
-    url "https://flintlib.org/downloads.html"
+    url "https://github.com/flintlib/flint/releases"
     regex(/href=.*?flint[._-]v?(\d+(?:\.\d+)+)\.t/i)
```
Comment (SMillerDev) [inline Formula/f/flint.rb:11]:
```suggestion
    url :stable
    strategy :github_latest
```

File: Formula/f/flint.rb (hunk 1 above)
Comment (edgarcosta) [reply Formula/f/flint.rb:45]: there is no make check. So I will remove the `--with-ntl`, as long it still installs the header

Comment (edgarcosta) [reply Formula/f/flint.rb]: I have confirmed that the headers are still being installed.

**Comment by edgarcosta** [comment]
@SMillerDev, thanks for the suggestion, I have also incorporated @albinahlback suggestion, but I still do not know how to address:
> Formulae should be verified as having support for runtime hardware detection before using ENV.runtime_cpu_detection.

**Comment by caarlos0** [comment]
> cc @caarlos0 for thoughts.

wrong cc maybe? I know nothing about flint or the underlying issue

**Comment by edgarcosta** [comment]
The underlying issue is the warning:
> Formulae should be verified as having support for runtime hardware detection before using ENV.runtime_cpu_detection

**Comment by edgarcosta** [comment]
I am honestly confused about the error message, 
```
No `cpuid` instruction detected. flint should not use `ENV.runtime_cpu_detection`.
```
as FLINT uses cpuid, see: https://github.com/flintlib/flint/blob/main/config/config.guess#L797

**Comment by albinahlback** [comment]
Have you taken a look at the [formula for GMP](https://github.com/Homebrew/homebrew-core/blob/master/Formula/g/gmp.rb)?

**Comment by cho-m** [comment]
> as FLINT uses cpuid, see: https://github.com/flintlib/flint/blob/main/config/config.guess#L797

Given that is config.guess, I think that is compile time usage of cpuid.

Runtime detection will result in code for multiple different microarchitectures being compiled into binary and then cpuid being used at runtime to select correct one to run. As an example, the binary may include AVX instructions, but a Core 2 Duo chip will still work by using the code path that only includes up to SSSE3 instructions.

If no cpuid is found within binaries, that usually means the binaries were compiled for a specific microarchitecture. These formulae should not use `ENV.runtime_cpu_detection` as they won't be compiled for general distribution to meet our minimum supported CPU (e.g. Core2/SSSE3 on Linux, Westmere/SSE4.2 on modern Intel macOS)

**Comment by mahrud** [comment]
Maybe you already know this, but you can locally run `brew style flint` to check that your formula passes the style and syntax checks before pushing. You can also do `brew style --fix flint` to autocorrect them in some cases.

**Comment by edgarcosta** [comment]
@mahrud Thank you!
I ended up learning it, perhaps a bit too late :)  it could have saved me a lot of time, as for a while I was struggling to see my obvious mistake.

**Comment by edgarcosta** [comment]
I think I fixed everything now, and this is ready to review

File: Formula/f/flint.rb [hunk 3]
```diff
@@ -21,25 +21,39 @@ class Flint < Formula
...
+  depends_on "automake" => :build
+  depends_on "libtool" => :build
   depends_on "gmp"
   depends_on "mpfr"
-  depends_on "ntl"
-
   uses_from_macos "m4" => :build
 
   def install
+    # to build against NTL
```
Comment (mahrud) [inline Formula/f/flint.rb:32]: NTL doesn't appear anywhere else in the formula now, is this correct?

File: Formula/f/flint.rb [hunk 4]
```diff
@@ -21,25 +21,39 @@ class Flint < Formula
...
+    if Hardware::CPU.intel?
+      # enable/disable avx{2,512}
+      # Because flint doesn't use CPUID at runtime
+      # we cannot rely on -march options
+      if build.bottle?
+        # prevent avx{2,512} in case we are building on a machine that supports it
+        args << "--enable-arch=#{Hardware.oldest_cpu}"
+      elsif Hardware::CPU.avx2?
+        # TODO: enable avx512 support
+        args << "--enable-avx2"
```
Comment (mahrud) [inline Formula/f/flint.rb:51]: Flint doesn't use BLAS yet, right? If it did, Apple's Accelerate framework has SIMD vector operations that replace avx: https://developer.apple.com/documentation/accelerate/simd

File: Formula/f/flint.rb (hunk 3 above)
Comment (albinahlback) [reply Formula/f/flint.rb:32]: Yes. We have a header-only interface.

File: Formula/f/flint.rb (hunk 4 above)
Comment (albinahlback) [reply Formula/f/flint.rb:51]: We do use it in multiplication for the `nmod_mat` and `fmpz_mat` modules.

File: Formula/f/flint.rb (hunk 3 above)
Comment (edgarcosta) [reply Formula/f/flint.rb:32]: and the NTL interface header is a C++ header

File: Formula/f/flint.rb (hunk 4 above)
Comment (edgarcosta) [reply Formula/f/flint.rb:51]: BLAS is optional; we could compile against it and add it as a dependency.
@albinahlback and @fredrik-johansson, is this something that we want?

**Review by carlocab — APPROVED** [review]
Thanks; great work!

**Comment by dimpase** [comment]
`pkg-config --modversion flint` returns 3.1.0 instead of 3.1.3.
For some reason the version is not bumped up, not sure it's here or upstream.

**Comment by cho-m** [comment]
> `pkg-config --modversion flint` returns 3.1.0 instead of 3.1.3. For some reason the version is not bumped up, not sure it's here or upstream.

Probably should open issue upstream for CI workflow relating to release tarballs to make sure `VERSION` file is modified before `bootstrap.sh` is run.

The problem looks like reverse is happening so `configure` contains:
```configure
PACKAGE_VERSION='3.1.0'
PACKAGE_STRING='FLINT 3.1.0'
```

https://github.com/flintlib/flint/blob/main/.github/workflows/release.yml#L88-L93
```yaml
      - name: "Bootstrap"
        run: |
          ./bootstrap.sh

      - name: "Create source archive"
        run: dev/make_dist.sh ${FLINT_VERSION}
```

https://github.com/flintlib/flint/blob/main/dev/make_dist.sh#L39-L40
```sh
# update VERSION file
printf $flint_version > VERSION
```

**Comment by edgarcosta** [comment]
I have submitted two PRs to fix the problem upstream: https://github.com/flintlib/flint/pull/1993 and https://github.com/flintlib/flint/pull/1994 (which FLINT releases if 3.1.4 will exist)

**Comment by carlocab** [comment]
Thanks! Feel free to open a PR that backports those patches to the current version of the formula.

**Pull request merged**